systemd: 249 -> 250

jonringer opened the pull request:
###### Motivation for this change
Continuation of #150491

Opening PR so I can start large build job before getting on a plane.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

andir committed:
systemd: 249.7 -> 250.3

andir committed:
systemd: reformat code with nixpkgs-fmt

andir committed:
systemdMinimal: don't set {libfido2,p11-kit,libgcrypt} to null

We don't have to do that as we already set all the feature flags to
null. Setting individual libraries to null instead of disabling their
feature flag will lead with bad example that will cause each of the
features to be disabled with multiple flags in the systemdMinimal
variant.

If a dependency is pulled in via another feature we should disable that
rather than setting it to null. Overriding a given package should be the
last resort.

andir committed:
systemd: remove unused lvm2 input

andir committed:
systemd: introduce withTests flag

This allows us to make test-only dependencies optional in builds that
aren't running tests (sadly all of our builds).

andir committed:
nixos/timesyncd: initialize clock file with current time

When initializing a system (e.g. first boot / livecd) we have no good
reference source for time. systemd-timesyncd however would revert back
to its configured fallback time (in our case 01.01.1980). Since we
probably don't want to hardcode a specific date as fallback we are now
using the current system time (wherever that might have come from) to
initialize the reference clock file.

The only systems that might be remotely affected by this change are
machines that have highly unreliable RTCs or those where the battery
that backs the RTC is running empty.

Historically these systems always had a tough time with anything time
related and likely required manual intervention.

For stateless systems (those that wipe / between reboots or our
installer CDs) this has the consequence that time will always be reset
to whatever the system comes up with on boot. This is likely the correct
time coming from an RTC. No harm done here the situation is likely
unchanged for them.

For stateful systems (those that retain the / partition across reboots)
there shouldn't be a change at all. They'll provide an initial clock
value once on their lifetime (during first boot / after installation).
From then onwards systemd-timesyncd will update the file with the newer
fallback time (that will be picked up on the next boot).

andir committed:
systemd: add the release timestamp into the build

This helps systemd during runtime to make decisions about the sanity of
the system clock. See the references news article for more details on
the matter.

flokli commented:
This needs to target staging, and needs to be bumped to systemd/systemd-stable v250.3.

Furthermore, we should at least skim over the changes from v250..v250.3  and verify that they don't break any NixOS things.

jonringer commented:
Yea, traveling right now. I'll mark as draft until it's actually ready

jonringer commented:
I'll do a proper rebase and merge conflict analysis once I get home.

jonringer commented:
@ofborg eval

flokli commented on nixos/doc/manual/release-notes/rl-2205.section.md:
````diff
@@ -59,6 +61,21 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 ## Backward Incompatibilities {#sec-release-22.05-incompatibilities}
 
+- Since systemd version 250 wireguard tunnels that are managed using networkd
+  will install routes for each of the `AllowedIPs` entries. This is in contrast
+  to the previous behaviour where routes had to be added manually.
+  To disable the new behaviour you must set the `RoutingTable=off` option in
+  the `WireGuard` section within each of your netdev definitions.
+  For example like this:
+  ```nix
+  {
+    systemd.network.netdev.my-wireguard-interface = {
+      wireguardConfig.RouteTable = "off";
+    };
+  }
````
This defaults flip been reverted in v250.3 (which this PR doesn't include yet).

@jonringer We should bump and revert this release note section.

jonringer replied:
it does contain 250.3

I squashed all of the bumps together

jonringer replied:
I will revert the release notes

jonringer replied:
reverted

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -123,7 +126,14 @@ assert withHomed -> withCryptsetup;
+  # command:
+  #  $ curl -s https://api.github.com/repos/systemd/systemd/releases/latest | \
+  #     jq '.created_at|strptime("%Y-%m-%dT%H:%M:%SZ")|mktime'
+  releaseTimestamp = "1640290180";
```
Can we use `$SOURCE_DATE_EPOCH` instead?

flokli replied:
It seems we can't, as we just fetch a git tag.  `SOURCE_DATE_EPOCH` is 315532800 (1980-01-01T00:00:00+00:00).

flokli replied:
Let's remove the commented-out

```
# "-Dtime-epoch=1"
```

from L480 though.

Izorkin commented:
@jonringer commit `systemd: 249.5 -> 250.3` needed rename to `systemd: 249.7 -> 250.3`

jonringer commented:
rebased on latest master

Mic92 commented:
I saw your PR too late and opened up https://github.com/NixOS/nixpkgs/pull/156438

Mic92 commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -489,65 +516,120 @@ stdenv.mkDerivation {
+            "src/analyze/test-verify.c"
+            "src/test/test-env-file.c"
+            "src/test/test-fileio.c"
+            "test/test-execute/exec-systemcallfilter-failing2.service"
```
I don't think we should attempt to patch tests.
I changed the code to replace tests instead: https://github.com/NixOS/nixpkgs/pull/156438/files#diff-a284772b505bd573f834b58591d380128545257d9e02601dbe112548526c8710R595-R597

Mic92 commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -415,7 +440,6 @@ stdenv.mkDerivation {
     "-Dnetworkd=${lib.boolToString withNetworkd}"
     "-Doomd=${lib.boolToString withOomd}"
     "-Dpolkit=${lib.boolToString withPolkit}"
-    "-Dcryptsetup=${lib.boolToString withCryptsetup}"
```
This option was just renamed and on staging we also use it:

```
"-Dlibcryptsetup=${lib.boolToString withCryptsetup}" 
```

Mic92 commented:
I am closing this just now to avoid pings on rebase...

Mic92 committed:
systemd: do not patch test files

We do not run them, so it is unnecessary work.

Mic92 commented:
```
systemctl --version
systemd 250 (250.3)
+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified
```

works on my laptop

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -487,65 +516,104 @@ stdenv.mkDerivation {
+      mkEnsureSubstituted = { replacement, search, where }:
+        ''
+          if [[ $(grep -r '${search}' | grep -v "${replacement}" | grep -Ev 'NEWS|^test/' | wc -l) -gt 0 ]]; then
+            echo "Not all references to '${search}' have been replace. Found the following matches:"
```
```suggestion
            echo "Not all references to '${search}' have been replaced. Found the following matches:"
```

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -612,6 +686,6 @@ stdenv.mkDerivation {
     platforms = platforms.linux;
     priority = 10;
-    maintainers = with maintainers; [ eelco flokli kloenk ];
+    maintainers = with maintainers; [ andir eelco flokli kloenk ];
```
31e5b8dc2198eacec3e009e21fc8226486aa674e

Mic92 replied:
added myself again.

jonringer commented:
I'll try this on my laptop as well.

I think we could probably merge into staging after next staging-next cycle begins

jonringer commented:
I'm running a hydra job on this: https://hydra.jonringer.us/jobset/nixpkgs/systemd250

flokli commented:
> I think we could probably merge into staging after next staging-next cycle begins

@jonringer there have been some more staging cycles (https://github.com/NixOS/nixpkgs/pull/158080, https://github.com/NixOS/nixpkgs/pull/158080).

This also needs a rebase, as there have been some conflicts. Is there anything else preventing this from being merged? (I see the PR title still contains a "WIP" too).

Mic92 commented:
I am using this for some time on my laptop now without noticeable impact so far.

jonringer commented:
@flokli I was going to use this staging cycle for glibc https://github.com/NixOS/nixpkgs/pull/133431#issuecomment-1053697179. Once it begins, I think we can merge this.

jonringer commented:
I can still rebase though

ius reviewed (changes requested):
Fails to build on EFI enabled RISC-V because the requested EFI linker (`efi-ld=gold`) is unsupported. According to Wikipedia gold only supports x86, x86-64, ARM, PowerPC, TileGX.

~~If there is no particular reason to pick gold over bfd, I suggest passing `-Defi-ld=bfd` instead.~~

Edit: I suggest just removing the option altogether, so it will figure out a suitable default by itself. See comment about changed semantics below as to why this is okay for v250.

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -435,7 +460,12 @@ stdenv.mkDerivation {
     "-Db_pie=true"
     "-Dinstall-sysconfdir=false"
-    "-Defi-ld=${stdenv.cc.bintools.targetPrefix}ld"
+    "-Defi-ld=gold"
```
Did this build before this PR?

ius replied:
Yes. In systemd v249.7 the variable was used to invoke the linker directly (see [meson.build](https://github.com/systemd/systemd-stable/blob/v249.7/src/boot/efi/meson.build#L276)), so that's why `ld` was valid and defaulted to `bfd`.

In systemd v250.3 it's passed to `-fuse-ld` instead.

mweinelt replied:
https://github.com/NixOS/nixpkgs/pull/150491#discussion_r767351535

flokli committed:
systemd: drop -Defi-ld=gold

As reported in
https://github.com/NixOS/nixpkgs/pull/156096#pullrequestreview-900986176,
this fails to build on EFI enabled RISC-V because the requested EFI
linker (efi-ld=gold) is unsupported. According to Wikipedia gold only
supports x86, x86-64, ARM, PowerPC, TileGX.

Removing this option alltogether will cause meson to figure out the
default linker by itself.

flokli commented:
I rebased this onto latest staging, fixed the conflicts, and dropped `-Defi-ld=gold` as suggested in https://github.com/NixOS/nixpkgs/pull/156096#discussion_r820131894.

Considering this has been open for quite some time, and tested by some people already, I'll run the `nixosTests.systemd{-boot}` tests once more, and then we should merge this.

flokli commented:
`nixosTests.systemd-boot` succeeded.

`nixosTests.systemd` couldn't run, due to `/nix/store/i3axx4kkdbfl01r0dfsps9qnj8biryxh-python3.9-scikit-build-0.12.0.drv` being broken right now. However, this has been sufficiently tested previously.

Let's get this into staging.

vcunat commented:
This merge broke `nixosTests.containers-ip`.  I checked the commit right before this merge and the merge commit (after some bisection steps, as I started with a range of many hundred commits).

Maybe also some other small-channel blockers from this list regressed together with it: https://hydra.nixos.org/eval/1748773#tabs-still-fail  (you can see logs there or reproduce cheaply on that particular commit)

flokli commented:
Thanks for the heads-up, I'll take a look.

flokli commented:
I built this locally. The container startup fails while trying to bind-mount `/nix/var/nix/daemon-socket` into the container:

```
machine # [    5.364468] systemd[1]: Starting Container 'webserver'...
machine # [    5.448067] container webserver[879]: Spawning container webserver on /var/lib/containers/webserver.
machine # [    5.448712] container webserver[879]: Press ^] three times within 1s to kill container.
machine # [    5.450500] container webserver[885]: Failed to stat /nix/var/nix/daemon-socket: No such file or directory
machine # [    5.454873] systemd[1]: container@webserver.service: Main process exited, code=exited, status=1/FAILURE
machine # [    5.455510] systemd[1]: container@webserver.service: Failed with result 'exit-code'.
machine # [    5.456098] systemd[1]: Failed to start Container 'webserver'.
machine # [    5.763319] systemd[1]: container@webserver.service: Scheduled restart job, restart counter is at 2.
machine # [    5.769093] systemd[1]: Stopped Container 'webserver'.
machine # [    5.770123] systemd[1]: Starting Container 'webserver'...
```

I ran the interactive test driver, and looked at the VM. Indeed `/nix/var/nix/daemon-socket` doesn't exist, because `nix-daemon.socket` failed to activate:

```
[  237.187747] systemd[1]: Nix Daemon Socket was skipped because of a failed condition check (ConditionPathIsReadWrite=/nix/var/nix/daemon-socket).
```

flokli commented:
Okay, I figured this out - this is actually a bug in the upstream nix-daemon.socket unit.

I have a workaround in nixpkgs, and will send a fix to the Nix repo.

flokli commented:
@vcunat the workaround is in https://github.com/NixOS/nixpkgs/pull/164644, and the proper fix in Nix is in https://github.com/NixOS/nix/pull/6282.
